pkg/fqdn/cache.go

```diff
@@ -154,16 +181,124 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
 		c.forward[entry.Name] = entries
 	}
 	c.updateWithEntryIPs(entries, entry)
+	if len(entries) > FQDNMaxIPperHost {
+		c.overLimit[entry.Name] = true
+	}
+
 	// When lookupTime is much earlier than time.Now(), we may not expire all
 	// entries that should be expired, leaving more work for .Lookup.
 	c.removeExpired(entries, time.Now(), time.Time{})
 }
 
+// AddNameToCleanup adds the IP with the given TTL to the the cleanup map to
+// delete the entry from the policy when it expires.
+// Need to be called with the lock
+func (c *DNSCache) addNameToCleanup(entry *cacheEntry) {
+	if entry.ExpirationTime.Before(c.lastCleanup) {
+		// ExpirationTime can be before the lastCleanup don't add that value to
+		// prevent leaks on the map.
+		return
+	}
+	expiration := entry.ExpirationTime.Unix()
+	expiredEntries, exists := c.cleanup[expiration]
+	if !exists {
+		expiredEntries = []string{}
+	}
+	c.cleanup[expiration] = append(expiredEntries, entry.Name)
+}
+
+// cleanupExpiredEntries cleans all the expired entries from the lastTime that
+// runs to the give time. It will lock the struct until retrieves all the data.
+// It returns the list of names that need to be deleted from the policies.
+func (c *DNSCache) cleanupExpiredEntries(expires time.Time) ([]string, time.Time) {
+	timediff := int(expires.Sub(c.lastCleanup).Seconds())
+	startTime := c.lastCleanup
+	expiredEntries := []string{}
+	for i := 0; i < int(timediff); i++ {
+		expiredTime := c.lastCleanup.Add(time.Second)
+		key := expiredTime.Unix()
+		c.lastCleanup = expiredTime
+		entries, exists := c.cleanup[key]
+		if !exists {
+			continue
+		}
+		expiredEntries = append(expiredEntries, entries...)
+		delete(c.cleanup, key)
+	}
+
+	result := []string{}
+	for _, name := range KeepUniqueNames(expiredEntries) {
+		if entries, exists := c.forward[name]; exists {
+			if !c.removeExpired(entries, c.lastCleanup, time.Time{}) {
+				continue
+			}
+			result = append(result, name)
+		}
+	}
+	return result, startTime
+}
+
+// cleanupOverLimitEntries returns the names that has reached the max number of
+// IP per host.
+// Internally the function sort the entries by the TTL.
```

Comment:
We should say sorted by expiration time. TTL might be ambiguous and mean the actual TTL seconds value.